patches/api/0449-Improve-Registry.patch

```diff
@@ -0,0 +1,117 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Wed, 20 Dec 2023 02:03:10 -0800
+Subject: [PATCH] Improve Registry
+
+Adds Registry#getKey(Object) which should be the
+primary way people get the key for an object. Registry
+items need to exist without having a key and so
+getKey() methods on Keyed objects that have a registry
+are marked as Deprecated or Obsolete.
+
+diff --git a/src/main/java/org/bukkit/Registry.java b/src/main/java/org/bukkit/Registry.java
+index 800d23bb249e19d5cf924e7ba36684068624da02..3fa25507e7af121b252bcd7521eca443283a6917 100644
+--- a/src/main/java/org/bukkit/Registry.java
++++ b/src/main/java/org/bukkit/Registry.java
+@@ -44,7 +44,9 @@ public interface Registry<T extends Keyed> extends Iterable<T> {
+      *
+      * @see Bukkit#getAdvancement(org.bukkit.NamespacedKey)
+      * @see Bukkit#advancementIterator()
++     * @deprecated use {@link Server#getAdvancement(NamespacedKey)} or {@link Server#advancementIterator()}
+      */
++    @Deprecated(since = "1.20.4") // Paper
+     Registry<Advancement> ADVANCEMENT = new Registry<Advancement>() {
+ 
+         @Nullable
+@@ -94,7 +96,9 @@ public interface Registry<T extends Keyed> extends Iterable<T> {
+      *
+      * @see Bukkit#getBossBar(org.bukkit.NamespacedKey)
+      * @see Bukkit#getBossBars()
++     * @deprecated use {@link Server#getBossBar(NamespacedKey)} or {@link Server#getBossBars()}
+      */
++    @Deprecated(since = "1.20.4") // Paper
+     Registry<KeyedBossBar> BOSS_BARS = new Registry<KeyedBossBar>() {
+ 
+         @Nullable
+@@ -143,7 +147,9 @@ public interface Registry<T extends Keyed> extends Iterable<T> {
+      * Default server loot tables.
+      *
+      * @see LootTables
++     * @deprecated use {@link LootTables} or {@link Server#getLootTable(NamespacedKey)}
+      */
++    @Deprecated(since = "1.20.4") // Paper
+     Registry<LootTables> LOOT_TABLES = new SimpleRegistry<>(LootTables.class);
+     /**
+      * Server materials.
+@@ -307,6 +313,22 @@ public interface Registry<T extends Keyed> extends Iterable<T> {
+     @Nullable
+     T get(@NotNull NamespacedKey key);
+ 
++    // Paper start - improve Registry
++    /**
++     * Get the key for this object.
```

Comment:
Should probably clarify here that getting null would be the exception, and give an example

Reply:
I'm also thinking there might be a use for a getKeyOrThrow method? Sometimes I hate nullable methods when I **know** it exists.